app/master/cluster_master.py

```diff
@@ -109,46 +109,24 @@ def connect_slave(self, slave_url, num_executors):
         """
         Connect a slave to this master.
 
-        If this slave has already been registered with the master and isn't working on a build, make sure that slave
-        is marked as alive and ready to accept builds.
-
         :type slave_url: str
         :type num_executors: int
         :return: The response with the slave id of the slave.
         :rtype: dict[str, str]
         """
+        # If a slave had previously been connected, and is now being reconnected, the cleanest way to resolve this
+        # bookkeeping is for the master to forget about the previous slave instance and start with a fresh instance.
         if slave_url in self._all_slaves_by_url:
-            self._logger.warning('Slave on {} requested to connect to master, even though already connected', slave_url)
-            slave = self._all_slaves_by_url.get(slave_url)
-            slave.set_is_alive(True)
-
-            # If the slave has made a request to reconnect to the master, it's safe to say that whatever
-            # subjob/build that slave was working on at the time is no longer running. In order to prevent
-            # that build from hanging indefinitely (as the slave will never respond to the master with the
-            # subjob results for that build) we must cancel the build.
-            if slave.current_build_id is not None:
-                self._logger.warning('Newly reconnected slave {} was working on build {}, cancelling build.',
-                                     slave_url, slave.current_build_id)
-                try:
-                    build = self.get_build(slave.current_build_id)
-                    build.cancel()
-                except ItemNotFoundError:
-                    pass
-
-            slave.current_build_id = None
-            # It's a safe operation to try to add the same slave to the slave allocator multiple
-            # times, as the underlying implementation is an OrderedSetQueue, which doesn't allow
-            # multiple entries.
-            self._slave_allocator.add_idle_slave(slave)
-            slave_id = slave.id
-        else:
-            slave = Slave(slave_url, num_executors)
-            self._all_slaves_by_url[slave_url] = slave
-            self._slave_allocator.add_idle_slave(slave)
-            self._logger.info('Slave on {} connected to master with {} executors. (id: {})',
-                              slave_url, num_executors, slave.id)
-            slave_id = slave.id
-        return {'slave_id': str(slave_id)}
+            self._logger.warning('Slave on {} requested to connect to master, even though previously connected. ' +
+                                 'Removing existing instance from master._all_slaves_by_url.', slave_url)
+            self._all_slaves_by_url.pop(slave_url)
```

Comment:
Joey and I had an offline conversation. We're going to keep what we have, but additionally we're going to perform a check on the old slave instance and see if it has a current_build_id set; we will then cancel it.